subprojects/core/src/integTest/groovy/org/gradle/api/internal/cache/CacheConfigurationsIntegrationTest.groovy

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.cache
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+
+class CacheConfigurationsIntegrationTest extends AbstractIntegrationSpec {
+
+    def "can configure caches via init script and query from settings script"() {
+        requireOwnGradleUserHomeDir()
+        def initDir = new File(executer.gradleUserHomeDir, "init.d")
+        initDir.mkdirs()
+        new File(initDir, "cache-settings.gradle") << """
+            beforeSettings { settings ->
+                settings.caches {
+                    ${modifiedCacheConfigurations}
+                }
+            }
+        """
+        settingsFile << """
+            caches {
+                assert releasedWrappers.removeUnusedEntriesAfterDays.get() == ${CacheConfigurationsInternal.DEFAULT_MAX_AGE_IN_DAYS_FOR_RELEASED_DISTS + 1}
+                assert snapshotWrappers.removeUnusedEntriesAfterDays.get() == ${CacheConfigurationsInternal.DEFAULT_MAX_AGE_IN_DAYS_FOR_SNAPSHOT_DISTS + 1}
+                assert downloadedResources.removeUnusedEntriesAfterDays.get() == ${CacheConfigurationsInternal.DEFAULT_MAX_AGE_IN_DAYS_FOR_DOWNLOADED_CACHE_ENTRIES + 1}
+                assert createdResources.removeUnusedEntriesAfterDays.get() == ${CacheConfigurationsInternal.DEFAULT_MAX_AGE_IN_DAYS_FOR_CREATED_CACHE_ENTRIES + 1}
+            }
+        """
+
+        expect:
+        succeeds("help")
+    }
+
+    def "cannot configure caches from settings script"() {
+        settingsFile << """
+            caches {
+                ${modifiedCacheConfigurations}
+            }
+        """
+
+        expect:
+        fails("help")
+        failureCauseContains("The value for property 'removeUnusedEntriesAfterDays' is final and cannot be changed any further")
```

Comment:
After reading the recently shared [error messages article](https://wix-ux.com/when-life-gives-you-lemons-write-better-error-messages-46c5223e1a2f), I can't help but see this as an example of a really terrible message to encounter.

It would be so much better if it could:

1. Talk about cache configuration in general, not just the specific property
2. Mention the solution, or at least link to a help page in our user guide
3. Not be a generic IllegalStateException, but be a more specific Gradle exception

As an experiment, I replaced the `DefaultSettings#caches` method with:

```
    @Override
    public void caches(Action<? super CacheConfigurations> cachesConfiguration) {
        try {
            cachesConfiguration.execute(getCaches());
        } catch (IllegalStateException e) {
            if (e.getMessage().contains("property 'removeUnusedEntriesAfterDays' is final"))
            throw new InvalidUserCodeException("You can not configure cache directories in a settings script, you must do this in an initscript, see http://userguide for more information.", e);
        }
    }
```

which then makes the failure output this:

```
* What went wrong:
A problem occurred evaluating settings 'trk7j'.
> You can not configure cache directories in a settings script, you must do this in an initscript, see http://userguide for more information.
   > The value for property 'removeUnusedEntriesAfterDays' is final and cannot be changed any further.
```

Which probably isn't how we want to do this, but produces (I think) a much more actionable error message.  I wish we had a more general purpose solution to allow us to easily add additional context to errors thrown when running configuration actions.

Reply:
This is obviously a deeper issue than just how we're using a finalized property here.  There really needs to be some context captured around why the value is finalized and what the user might be trying to do when they trigger an error, neither of which do properties currently capture.  I don't think moving away from a property here would be the right thing to do as much of its behavior is required and it would be unfortunate to duplicate it.  Although I wholeheartedly agree with the sentiment that this error could be much better, I'm not inclined to do anything about this now, but we should revisit this if we ever have a better way to record this sort of context for properties in the future.